File: scripts/integrations/package_for_download.py
#!/usr/bin/env python3
"""
package_for_download.py

PURPOSE: Package completed integration for user download

USAGE: python package_for_download.py <project-directory> [version-name]

PARAMETERS:
  project-directory - Path to integration project directory
  version-name - (Optional) Version name (e.g., "v1", "final")

EXIT CODES:
  0 - Success: Package created
  1 - Error: Project directory not found
  2 - Error: Zip creation failed

"""

import os
import sys
import zipfile
from datetime import datetime


def create_package(project_dir, version_name=None):
    """Create zip package of integration."""
    if not os.path.isdir(project_dir):
        print(f"❌ Project directory not found: {project_dir}")
        return 1

    project_name = os.path.basename(os.path.abspath(project_dir))
    if version_name:
        package_name = f"{project_name}-{version_name}.zip"
    else:
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        package_name = f"{project_name}-{timestamp}.zip"

    outputs_dir = "/mnt/user-data/outputs"
    os.makedirs(outputs_dir, exist_ok=True)
    output_path = os.path.join(outputs_dir, package_name)

    print(f"📦 Packaging integration: {project_dir}")
    print(f"Package: {package_name}")
    print("")

    try:
        file_count = 0
        excluded_patterns = [
            "node_modules",
            ".git",
            "__pycache__",
            ".DS_Store",
            ".env",
            "components",
        ]

        with zipfile.ZipFile(output_path, "w", zipfile.ZIP_DEFLATED) as zipf:
            for root, dirs, files in os.walk(project_dir):
                dirs[:] = [d for d in dirs if d not in excluded_patterns]

                rel_root = os.path.relpath(root, os.path.dirname(project_dir))

                should_skip = False
                for pattern in excluded_patterns:
                    if pattern in root:
                        should_skip = True
                        break

                if should_skip:
                    continue

                for file in files:
                    if file in excluded_patterns or file.endswith(".pyc"):
                        continue

                    file_path = os.path.join(root, file)
                    arcname = os.path.join(rel_root, file)

                    zipf.write(file_path, arcname)
                    file_count += 1

        size_bytes = os.path.getsize(output_path)
        size_kb = size_bytes / 1024
        size_mb = size_kb / 1024

        print("✅ Package created successfully")
        print("")
        print(f"Files included: {file_count}")
        if size_mb >= 1:
            print(f"Package size: {size_mb:.2f} MB")
        else:
            print(f"Package size: {size_kb:.2f} KB")
        print("")
        print(f"Download location: {output_path}")
        print(f"Download link: computer:///{output_path}")
        print("")
        print("📋 Package contents:")
        print("  - Source code (src/)")
        print("  - Built artifacts (dist/)")
        print("  - Configuration files (package.json, tsconfig.json)")
        print("  - Documentation")
        print("")
        print("ℹ️  Excluded from package:")
        print("  - node_modules/ (dependencies)")
        print("  - components/ (downloaded component source)")
        print("  - .git/ (version control)")
        print("  - .env (secrets)")
        print("")
        print("To use this package:")
        print("  1. Download the zip file")
        print("  2. Extract on your local machine")
        print("  3. Run: npm install")
        print("  4. Deploy to Prismatic")

        return 0

    except Exception as e:
        print(f"❌ Error creating package: {e}")
        print("")
        print(f"Error type: {type(e).__name__}")
        print(f"Error message: {str(e)}")
        return 2


def main():
    if len(sys.argv) < 2:
        print("❌ No project directory provided")
        print(
            "Usage: python package_for_download.py <project-directory> [version-name]"
        )
        return 1

    project_dir = sys.argv[1]
    version_name = sys.argv[2] if len(sys.argv) > 2 else None

    return create_package(project_dir, version_name)


if __name__ == "__main__":
    sys.exit(main())

File: plugin/scripts/integrations/package-for-download.ts
#!/usr/bin/env node
/**
 * package-for-download.ts
 *
 * PURPOSE: Package completed integration for user download
 *
 * USAGE: node package-for-download.ts <project-directory> [version-name]
 *
 * EXIT CODES:
 *   0 - Success: Package created
 *   1 - Error: Project directory not found
 *   2 - Error: Zip creation failed
 */

import type { Dirent } from "node:fs";
import { mkdir, readdir, stat } from "node:fs/promises";
import { join, resolve, basename, relative } from "node:path";
import { tmpdir } from "node:os";
import { exec } from "../../vendor/tinyexec/main.mjs";
import { confineToProjectRoot } from "../shared/project-directory.ts";

const EXCLUDED_PATTERNS = [
  "node_modules",
  ".git",
  "__pycache__",
  ".DS_Store",
  ".env",
  "components",
];

const walkDir = async (dir: string, baseDir: string, excluded: string[]): Promise<string[]> => {
  const results: string[] = [];

  let entries: Dirent[];
  try {
    entries = await readdir(dir, { withFileTypes: true });
  } catch {
    return results;
  }

  for (const entry of entries) {
    if (excluded.includes(entry.name)) continue;

    const fullPath = join(dir, entry.name);

    // Check if any parent path component is excluded
    const relPath = relative(baseDir, fullPath);
    const parts = relPath.split(/[/\\]/);
    if (parts.some((p) => excluded.includes(p))) continue;

    if (entry.isDirectory()) {
      results.push(...(await walkDir(fullPath, baseDir, excluded)));
    } else if (entry.isFile() && !entry.name.endsWith(".pyc")) {
      results.push(fullPath);
    }
  }

  return results;
};

const createPackage = async (projectDir: string, versionName?: string): Promise<number> => {
  try {
    if (!(await stat(projectDir)).isDirectory()) {
      console.log(`Project directory not found: ${projectDir}`);
      return 1;
    }
  } catch {
    console.log(`Project directory not found: ${projectDir}`);
    return 1;
  }

  const projectName = basename(resolve(projectDir));
  let packageName: string;
  if (versionName) {
    packageName = `${projectName}-${versionName}.zip`;
  } else {
    const now = new Date();
    const timestamp = [
      now.getFullYear(),
      String(now.getMonth() + 1).padStart(2, "0"),
      String(now.getDate()).padStart(2, "0"),
      "-",
      String(now.getHours()).padStart(2, "0"),
      String(now.getMinutes()).padStart(2, "0"),
      String(now.getSeconds()).padStart(2, "0"),
    ].join("");
    packageName = `${projectName}-${timestamp}.zip`;
  }

  const outputsDir =
    process.env.PRISMATIC_OUTPUT_DIR ||
    join(process.env.HOME || process.env.USERPROFILE || tmpdir(), "prismatic-outputs");
  await mkdir(outputsDir, { recursive: true });
  const outputPath = join(outputsDir, packageName);

  console.log(`Packaging integration: ${projectDir}`);
  console.log(`Package: ${packageName}`);
  console.log("");

  try {
    // Use the zip command which is available on most systems
    const files = await walkDir(projectDir, resolve(projectDir, ".."), EXCLUDED_PATTERNS);

    if (files.length === 0) {
      console.log("No files to package");
      return 2;
    }

    // Build relative paths for zip
    const parentDir = resolve(projectDir, "..");
    const relFiles = files.map((f) => relative(parentDir, f));

    // Use system zip command
    let zipError = "";
    let zipCreated = false;
    try {
      const result = await exec("zip", ["-r", outputPath, ...relFiles], {
        timeout: 60000,
        nodeOptions: { cwd: parentDir },
      });
      zipCreated = result.exitCode === 0;
      zipError = result.stderr;
    } catch (error) {
      zipError = String(error);
    }

    if (!zipCreated) {
      // Fallback: try tar.gz if zip is not available
      const tarPath = outputPath.replace(".zip", ".tar.gz");
      let tarError = "";
      let tarCreated = false;
      try {
        const tarResult = await exec("tar", ["-czf", tarPath, "-C", parentDir, ...relFiles], {
          timeout: 60000,
        });
        tarCreated = tarResult.exitCode === 0;
        tarError = tarResult.stderr;
      } catch (error) {
        tarError = String(error);
      }

      if (!tarCreated) {
        console.log("Error creating package");
        console.log((tarError || zipError).slice(0, 500));
        return 2;
      }

      const archiveStat = await stat(tarPath);
      const sizeKb = archiveStat.size / 1024;
      const sizeMb = sizeKb / 1024;

      console.log("Package created successfully (tar.gz)");
      console.log("");
      console.log(`Files included: ${files.length}`);
      console.log(
        `Package size: ${sizeMb >= 1 ? `${sizeMb.toFixed(2)} MB` : `${sizeKb.toFixed(2)} KB`}`,
      );
      console.log("");
      console.log(`Download location: ${tarPath}`);
      return 0;
    }

    const archiveStat = await stat(outputPath);
    const sizeKb = archiveStat.size / 1024;
    const sizeMb = sizeKb / 1024;

    console.log("Package created successfully");
    console.log("");
    console.log(`Files included: ${files.length}`);
    console.log(
      `Package size: ${sizeMb >= 1 ? `${sizeMb.toFixed(2)} MB` : `${sizeKb.toFixed(2)} KB`}`,
    );
    console.log("");
    console.log(`Download location: ${outputPath}`);
    console.log(`Download link: computer:///${outputPath}`);
    console.log("");
    console.log("Package contents:");
    console.log("  - Source code (src/)");
    console.log("  - Built artifacts (dist/)");
    console.log("  - Configuration files (package.json, tsconfig.json)");
    console.log("  - Documentation");
    console.log("");
    console.log("Excluded from package:");
    console.log("  - node_modules/ (dependencies)");
    console.log("  - components/ (downloaded component source)");
    console.log("  - .git/ (version control)");
    console.log("  - .env (secrets)");
    console.log("");
    console.log("To use this package:");
    console.log("  1. Download the zip file");
    console.log("  2. Extract on your local machine");
    console.log("  3. Run: npm install");
    console.log("  4. Deploy to Prismatic");

    return 0;
  } catch (e) {
    console.log(`Error creating package: ${e}`);
    return 2;
  }
};

const main = async (): Promise<number> => {
  if (process.argv.length < 3) {
    console.log("No project directory provided");
    console.log("Usage: node package-for-download.ts <project-directory> [version-name]");
    return 1;
  }

  let projectDir: string;
  try {
    projectDir = confineToProjectRoot(process.argv[2]);
  } catch (e) {
    console.log((e as Error).message);
    return 1;
  }
  const versionName = process.argv[3] ?? undefined;

  return createPackage(projectDir, versionName);
};

process.exit(await main());
